pkg/learning, syz-fuzzer: add MAB-based seed scheduling

SoveraNia opened the pull request:
Detailed changes:
    pkg/learning: core MAB seed scheduling logic.
    syz-fuzzer: modify existing tasks to collect coverage and timing information.

SoveraNia committed:
pkg/learning: add Multi-armed-bandit engine

To be later used for workflow choices (e.g. seed selection, task selection, etc.).

SoveraNia committed:
syzkaller: add MAB-based seed scheduling

Detailed changes:
pkg/learning: core MAB seed scheduling logic.
syz-fuzzer, pkg/ipc: modify existing tasks to collect coverage and timing information.

dvyukov commented:
Sorry for delays, I am now overloaded with the remote intern season. Looking at the PR now.

dvyukov commented on syz-fuzzer/proc.go:
```diff
@@ -89,6 +90,20 @@ func (proc *Proc) loop() {
+			// There is no need to use snapshot here as MAB engine has its own lock.
+			// Also, the size of proc.fuzzer.corpus can only increase.
+			pidx, pr := proc.fuzzer.mabSS.Choose(proc.rnd)
+			if pidx < 0 || pidx >= len(proc.fuzzer.corpus) {
```
We already checked that corpus is not empty, so pidx < 0 also looks impossible/bug.

SoveraNia replied:
Removed.

dvyukov commented on pkg/learning/exec_result.go:
```diff
@@ -0,0 +1,9 @@
+
+package learning
+
+type ExecResult struct {
```
This is reasonable to move to mab_ss.go because it's only used for MABSeedScheduler and is specific to MABSeedScheduler. File per type is generally excessive.

SoveraNia replied:
I originally put it here so the planned TaskScheduler can use it as well. But perhaps it's better to move it to SeedScheduler and move it out later if needed.

dvyukov commented on syz-fuzzer/proc.go:
```diff
@@ -247,16 +267,23 @@ func (proc *Proc) executeHintSeed(p *prog.Prog, call int) {
-	calls, extra := proc.fuzzer.checkNewSignal(p, info)
+func (proc *Proc) execute(execOpts *ipc.ExecOpts, p *prog.Prog, flags ProgTypes,
+	stat Stat) (*ipc.ProgInfo, learning.ExecResult) {
+	ret := learning.ExecResult{
```
`var ret learning.ExecResult`

SoveraNia replied:
Done with:
```
ret := learning.ExecResult{
    Coverage: cov,
    ExecTime: time,
}
```

dvyukov commented on syz-fuzzer/proc.go:
```diff
@@ -247,16 +267,23 @@ func (proc *Proc) executeHintSeed(p *prog.Prog, call int) {
+	info, time := proc.executeRaw(execOpts, p, stat)
+	ret.TimeExec = time
+	calls, extra, cov := proc.fuzzer.checkNewSignal(p, info)
+	ret.Coverage = cov
```
Or better move ret initialization to here and skip the intermediate hops:
```
ret := learning.ExecResult{
    Coverage: cov,
    ExecTime: time,
}
```

SoveraNia replied:
Done,

SoveraNia commented on pkg/learning/mab.go:
The MAB algorithm is designed to use probability when the choice is made. Update, however, is called after making choice, mutation and executing the program, at which point other procs could potentially update the engine to a new pr.

SoveraNia commented on pkg/learning/mab.go:
The check is to prevent overflow as well as messing up with probability computation. Given how the algorithm works it would take a very very long time for the reward to actually reach this threshold.

Come to think of it, this should probably be THRESHOLD / numChoices, since the largest value used in this function is sumWeights.

dvyukov replied:
I see it's used multiple times, please define it on the package level with a comment as:
```
	var weightThreshold = math.Exp(16)
```
or:
```
	const weightThreshold = 1e10
```

SoveraNia commented on pkg/learning/mab.go:
Yes, since weight = Exp(reward) it's much more prune to overflowing. However, if one single choice has a significantly larger reward, we still want to keep track of it.

It might be possible that we get into a situation where all choices as overflowing weights. The solution is proposed in the previous comment.

SoveraNia commented on pkg/learning/mab.go:
```diff
@@ -0,0 +1,109 @@
+	return idx, mab.Choices[idx].weight / sumWeights
+}
+
+func (mab *MultiArmedBandit) NewChoice() int {
```
It's actually used in the test where the index of the new choice is checked. Admittedly, we could let the caller compute the index themselves by computing length-1. We might need additional interfaces for that as choices array are now private.

dvyukov replied:
I see it is used in 1 test as follows:
```
	for c := 0; c < numChoices; c++ {
		ret := mab.NewChoice()
		if ret != c {
			t.Fatalf("error adding new choice %v. got return value %v", c, ret)
		}
		... use mab.choices[ret] ...
```

The test is in the same package and accesses mab.choices and assumes the returned value is the index anyways. So it can be rewritten as follows w/o losing much of the checking (or may be making it better, as we now also check exact mab.choices size):

```
	for c := 0; c < numChoices; c++ {
		mab.NewChoice()
		if len(mab.choices) != c + 1 {
			t.Fatalf("error adding new choice %v. got return value %v", c, ret)
		}
		... use mab.choices[c] ...
```

SoveraNia commented on pkg/learning/mab.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+type MultiArmedBandit struct {
+	Mu      sync.RWMutex
```
Got it, changing them in the upcoming amend.

SoveraNia replied:
Done.

SoveraNia commented on pkg/learning/mab.go:
```diff
@@ -0,0 +1,109 @@
+		return -1, 0.0
+	}
+	sumWeights := mab.Choices[len(mab.Choices)-1].sumWeights
+	randVal := r.Float64() * mab.Choices[len(mab.Choices)-1].sumWeights
```
Thanks for catching. Working on it in the amend.

SoveraNia commented on pkg/learning/mab.go:
```diff
@@ -0,0 +1,109 @@
+	if len(mab.Choices) > 0 {
+		newChoice.sumWeights = mab.Choices[len(mab.Choices)-1].sumWeights + newChoice.weight
+	}
+	mab.Choices = append(mab.Choices, newChoice)
```
This function should be protected by mutex. Thanks for the testing suggestion.

SoveraNia commented on pkg/learning/mab.go:
This sounds reasonable. Will change these to log.Fatalf for all unexpected inputs.

A question: what's the best practice to test these fatal conditions? The solutions I found seems a bit hacky to me.

dvyukov replied:
The proper way is to return an error, then callers can handle it in a suitable way: crash, log and continue, check in the test, etc.

SoveraNia commented on pkg/learning/mab.go:
I don't think this is a good idea as that will require the indices (which is used in more places) to be shifted by 1, which increases code complexity and could introduce bugs.

SoveraNia commented on syz-fuzzer/proc.go:
```diff
@@ -201,12 +216,17 @@ func getSignalAndCover(p *prog.Prog, info *ipc.ProgInfo, call int) (signal.Signa
 	if proc.fuzzer.comparisonTracingEnabled && item.call != -1 {
 		proc.executeHintSeed(item.p, item.call)
 	}
+	// If MAB seed scheduling enabled, do not mutate for 100 times.
```
Yes, in the research we disabled all smash mutations when MAB is in play. In practice, the 100 forced smashes will cause an extremely long smash-triage-smash chain reaction, which is actually detrimental to corpus diversity. With MAB, old seeds are very likely to have negative weight and new seeds will still be mutated.

Of course, it can be argued that we could reduce the number of smashes performed. In the research, we went for the extreme to demonstrate MAB can do it just as well. 

Also, I realize that if we do want to disable smash mutations completely, the fault inject logic will also need to be moved as well.

SoveraNia replied:
I had an idea that might get us the best of both worlds: Run the smash mutations for up to 100 times and compute their reward as if they were chosen with a certain probability (e.g. 100%). Break the smash loop as soon as the total seed reward drops below a certain threshold (e.g. 0). This way we can thoroughly test the new seeds first (which can be slow for MAB alone when there are thousands of seeds) while abandoning bad seeds quickly.

Needs further testing.

dvyukov replied:
> Break the smash loop as soon as the total seed reward drops below a certain threshold (e.g. 0).

This sounds reasonable.
But one way or another dead code is not good, especially in such covert form. The code that never runs should be removed.

SoveraNia commented on syz-fuzzer/fuzzer.go:
```diff
@@ -488,30 +498,38 @@ func (fuzzer *Fuzzer) corpusSignalDiff(sign signal.Signal) signal.Signal {
 	fuzzer.signalMu.Unlock()
 	fuzzer.signalMu.RLock()
-	return true
+	return len(diff)
```
I designed this to return the new coverage achieved, which depends on what's already in the corpus.

SoveraNia commented on pkg/learning/mab_ss.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (ss *MABSeedScheduler) Update(idx int, result ExecResult, pr float64) {
+	ss.totalTime += result.TimeExec
```
Added mutex guarding the Update function.

SoveraNia commented on syz-fuzzer/fuzzer.go:
```diff
@@ -60,6 +61,10 @@ type Fuzzer struct {
 	logMu sync.Mutex
+
+	// Multi-armed-bandit seed selection
+	mabSSEnabled bool
```
I'm wondering whether we should have an option in pkg/learning/mgrconfig that turns the feature off should the user desire for any reason, or have some easy way to disable this like the corpus rotation feature. Especially considering for MabSS to work best the Smash workflow will need to change. What do you think?

dvyukov replied:
I am not a fan of such tunables.
These decisions is the hardest part of the problem. If you cannot decide should it be enabled or not, nobody can decide. Most syzkaller users are totally uninterested in such details and don't have any relevant expertise, nor desire to spend time on these things.
Should we enable or disable it on syzbot? No idea. But whatever we chose, we can have only 1 setting and the other one becomes irrelevant.
I could expose a thousand of such bool/integer tunables from the fuzzer engine, but then nobody besides me could run syzkaller efficiently.

SoveraNia commented on syz-fuzzer/proc.go:
```diff
@@ -89,6 +90,16 @@ func (proc *Proc) loop() {
+			// Use MAB engine to pick a prog.
+			// There is no need to use snapshot here as MAB engine has its own lock.
+			// Also, the size of proc.fuzzer.corpus can only increase.
+			pidx, pr := fuzzerSnapshot.mabSS.Choose(proc.rnd)
```
I can see there is a concurrency problem here: fuzzerSnapshot and mabSS.Choose is protected by different mutexes and it's possible for mabSS's status to change (since it's a pointer) after fuzzerSnapshot is taken. We can either
1) Let fuzzerSnapshot has a deep copy of mabSS (including its MAB engine) or
2) Let mabSS maintain its own copy of seed corpus array.

Any suggestions?

SoveraNia replied:
I have modified the seed selection engine to maintain its own corpus. This way we don't have to create snapshots in two different places and risk race conditions. It's also helpful for the synchronization process since the fuzzer needs to access the seed program for synchronization.

dvyukov replied:
What exactly is the problem?
I see that we need the chosen program index pidx to be not larger than the corpus size. This can be ensured by taking snapshot after selecting a program I think.
The corpus passed to p.Mutate does not need to be consistent with anything.

SoveraNia commented on syz-fuzzer/proc.go:
```diff
@@ -301,8 +329,9 @@ func (proc *Proc) executeRaw(opts *ipc.ExecOpts, p *prog.Prog, stat Stat) *ipc.P
 		}
-		log.Logf(2, "result hanged=%v: %s", hanged, output)
-		return info
+		tsEnd := time.Now().UnixNano()
```
Would IPC still be able to measure the time correctly if some syscall changes the OS timers in the guest VM?

In the research, we created an upper/lower bound to filter out abnormal time measurements. This is hacky and not good for the long-term sanity of the MAB engine.

SoveraNia replied:
I migrated the timing measurement to IPC package. Didn't address the OS time change issue so I give executions with bad timing measurements special treatment.

Need to test how that work. According to my experiments handling time differently could impact how the learning algorithm determines which seed to favor.

dvyukov replied:
> Would IPC still be able to measure the time correctly if some syscall changes the OS timers in the guest VM?

This code also runs inside of the VM, so wouldn't it be equally affected by time changes?
But I think UnixNano gives monotonic clock time, so should not be subject to changes.

SoveraNia commented on syz-fuzzer/proc.go:
```diff
@@ -89,6 +90,20 @@ func (proc *Proc) loop() {
+				log.Logf(1, "#%v: seed program index (%v) out-of-bound. skipping", proc.pid, pidx)
+				continue
+			}
+			p := proc.fuzzer.corpus[pidx]
```
I have modified the seed selection engine to maintain its own corpus. This way we don't have to create snapshots in two different places and risk race conditions.

SoveraNia commented on pkg/learning/mab_ss.go:
```diff
@@ -33,12 +34,15 @@ type MABSeedScheduler struct {
 	// index in the MAB engine.
 	corpus []*prog.Prog
+	// Reward change since last Poll.
+	rewardChange map[int]float64
```
I really want to sync changes to totalTime and totalCov as well and restore them after fuzzer crash-and-restarts. We did that in the research with a single VM. However, that wouldn't make sense (especially the totalTime part) if different fuzzers are running on VMs with different computational powers.

Is this common in practice?

dvyukov replied:
We need to try to prototype #1541 and if it works, then it will make syncing unnecessary.

SoveraNia replied:
Folding the fuzzer into the manager could definitely reduce the syncing overhead. Still, the timing measurement could be affected if there is a discrepancy between VM performances. 

Are there any solid attempts underway to making #1541 happen?

dvyukov replied:
The VMs are generally assumed to be equivalent.

> Are there any solid attempts underway to making #1541 happen?

None that I know of.

SoveraNia committed:
pkg/learning: apply reward adjustment on new choices

If a new choice is added after reward adjustment is mad by the MAB engine,
adjust the reward of the new choice correctly.

dvyukov commented on pkg/learning/mab.go:
This looks fishy. Did mask data races when mutex was omitted in some places?
Who is calling this with wrong idx and why?

dvyukov commented on syz-manager/rpc.go:
```diff
@@ -302,6 +314,23 @@ func (serv *RPCServer) Poll(a *rpctype.PollArgs, r *rpctype.PollRes) error {
+		serv.mabSSRewards[sig] += reward
+		log.Logf(4, "MAB SS reward update %v: %v", sig.String(), reward)
+	}
+	// TODO(Daimeng): Send server-side reward update to fuzzer, when there are
```
The case of multiple fuzzers is very important. In most real uses there are multiple fuzzers, sometimes I use up to 40. What will it take to resolve this TODO?
